### PR TITLE
Jf/74

### PR DESCRIPTION
**Pull request by gnulnx**

### CODE REVIEW

File: mongolog/handlers.py [hunk 1]
```diff
@@ -231,14 +154,40 @@ def check_keys(self, record):
...
-            record['msg'][self.new_key(k)] = record['msg'].pop(k)
-
-            if isinstance(v, dict):
-                for old_key in record['msg'][k].keys():
-                    record['msg'][k][self.new_key(old_key)] = record['msg'][k].pop(old_key)
+            self._check_keys(k, v, record['msg'])
 
         return record
 
+    def _check_keys(self, k, v, _dict):
```
Comment (gnulnx) [inline mongolog/handlers.py]: This is now a recursive check that goes as deep as required to check all key's.  Previously it only went 2 levels deep.

File: mongolog/handlers.py [hunk 2]
```diff
@@ -231,14 +154,40 @@ def check_keys(self, record):
...
+        """
+        Recursively check key's for special characters.
+        """
+        _dict[self.new_key(k)] = _dict.pop(k)
+
+        if isinstance(v, dict):
+            for nk, vk in v.items():
+                self._check_keys(nk, vk, v)
```
Comment (codekiln) [inline mongolog/handlers.py]: Does this need another branch to check if it's a list containing objects with further keys?

File: mongolog/handlers.py [hunk 3]
```diff
@@ -278,6 +227,22 @@ def create_log_record(self, record):
...
+
+        This default implementation just uses
+        traceback.print_exception()
+        """
+        sio = StringIO()
+
+        traceback.print_exception(ei[0], ei[1], ei[2], None, sio)
+        s = sio.getvalue()
+        sio.close()
+        if s[-1:] == "\n":
```
Comment (codekiln) [inline mongolog/handlers.py]: Perhaps use [`os.linesep`](https://docs.python.org/3/library/os.html#os.linesep) for those poor souls on windows

File: mongolog/handlers.py [hunk 4]
```diff
@@ -231,14 +154,40 @@ def check_keys(self, record):
...
+            As of mongo 3.6 . and $ are permitted in keys.  But keys may not start with $
+            If we encounter a key that starts with a $ we replace it with it's unicode full width equivalent.
+        """
+        if self.mongo_version >= 3.6:
+            if key[0] == "$":
+                key = u"＄" + key[1:]
+        else:
+            # Older mongo doesn't support $ or . in keys
+            if "." in key:
+                key = key.replace(u".", u"．")
```
Comment (codekiln) [inline mongolog/handlers.py]: Crazy, I've never seen `u"．"` before.

Comment (gnulnx) [reply mongolog/handlers.py]: These are the full width unicode equivalents... This was a recommended approach.  Normally I'd rather through and error to let user know that there is an invalid key... but when it comes to logging you can't really afford to ever let it fail so the special characters are swapped out for the full width equivalents.

File: mongolog/handlers.py (hunk 2 above)
Comment (gnulnx) [reply mongolog/handlers.py]: Yep you are right.  Just did some quick testing and it fails with that currently.  I'll update and push changes a bit later.

File: mongolog/handlers.py [hunk 5]
```diff
@@ -327,10 +291,10 @@ def insert_embedded(self, log_record):
             # First time this record has been seen add a created field and insert it
             log_record['created'] = log_record.pop('time')
             log_record['counter'] = 1
-            if pymongo_version == 2:
-                self.mongolog.insert(log_record)
-            elif pymongo_version == 3:
+            if pymongo_version >= 3:
                 self.mongolog.insert_one(log_record)
+            elif pymongo_version == 2:
+                self.mongolog.insert(log_record)
```
Comment (codekiln) [inline mongolog/handlers.py]: Not sure if this is necessary given setup.py, but worth considering here:

```
else:
    raise UnsupportedVersionError("Mongolog only supports versions >= 2")
```

**Pull request merged**